csrc/accp_lc_loader.cpp

```diff
@@ -0,0 +1,45 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <dlfcn.h>
+#include "generated-headers.h"
+#include "env.h"
+
+using namespace AmazonCorrettoCryptoProvider;
+
+// The function in the symbol table we use to recognize that libcrypto has been loaded.
+#define CANARY_FUNCTION_NAME STRINGIFY(AWSLC_MANGLE) "CRYPTO_library_init"
+
+JNIEXPORT jboolean JNICALL Java_com_amazon_corretto_crypto_provider_Loader_loadLibCrypto(
+    JNIEnv* pEnv, jclass, jstring libPath)
+{
+    try {
+        raii_env env(pEnv);
+
+        // Have we already loaded a properly mangled version of AWS-LC?
+        dlerror(); // Clear errors
+        dlsym(RTLD_DEFAULT, CANARY_FUNCTION_NAME); // Return value doesn't matter
+        char* dlErrorReturned = dlerror(); // Did the lookup succeed?
```

Comment:
if i understand this correctly, this block will tell us whether _some library_ presenting the symbol `CRYPTO_library_init` has been loaded by the process, but not necessarily that it's the exact library we desire (AWS-LC and BoringSSL export many identical symbol names). perhaps we want to be more precise in our checks? it  looks like we can use `dlopen` with the `RTLD_NOLOAD` flag set to test whether the specific library we want has already been loaded:

> RTLD_NOLOAD (since glibc 2.2)
Don't load the library. This can be used to test if the library is already resident (dlopen() returns NULL if it is not, or the library's handle if it is resident). This flag can also be used to promote the flags on a library that is already loaded. For example, a library that was previously loaded with RTLD_LOCAL can be reopened with RTLD_NOLOAD | RTLD_GLOBAL. This flag is not specified in POSIX.1-2001.

https://linux.die.net/man/3/dlerror